### PR TITLE
Properly initialize the Boehm GC on all platforms.

### PR DESCRIPTION
**Pull request by rbehrends**
Initialize Boehm GC correctly on all platforms.

### CODE REVIEW

**Comment by rbehrends** [comment]
Also register threads properly.

File: lib/system/threads.nim [hunk 1]
```diff
@@ -135,10 +135,18 @@ else:
...
-            header: pthreadh.}
+  when not defined(boehmgc):
+    proc pthread_create(a1: var SysThread, a2: var PthreadAttr,
+              a3: proc (x: pointer): pointer {.noconv.},
+              a4: pointer): cint {.importc: "pthread_create",
+              header: pthreadh.}
+  else:
+    # The Boehm GC needs to know about newly created threads,
+    # so pthread_create() is redirected to GC_pthread_create().
+    proc pthread_create(a1: var SysThread, a2: var PthreadAttr,
```
Comment (Araq) [inline lib/system/threads.nim]: That's ok, but on Windows we don't call pthread_create at all.

File: lib/system.nim [hunk 2]
```diff
@@ -1200,6 +1200,14 @@ const
   hasSharedHeap = defined(boehmgc) or defined(gogc) # don't share heaps; every thread has its own
   taintMode = compileOption("taintmode")
 
+when defined(boehmgc):
```
Comment (Araq) [inline lib/system.nim]: I don't like it in system but I suspect you tried and failed to put it in system/threads.nim.

Comment (rbehrends) [inline lib/system.nim]: The problem is that `system/threads.nim` gets included before `system/mmdisp.nim`, and if I put it in `system/threads.nim`, it won't be available if `--threads:off` is selected. I also thought about putting it in a separate include file, but that would be wasteful. I'm open for an alternative suggestion; reordering the includes in `system.nim` seems to invariably break stuff.

File: lib/system/threads.nim (hunk 1 above)
Comment (rbehrends) [inline lib/system/threads.nim]: Okay, I'll restructure the PR to avoid a `pthread_create()` wrapper.

File: lib/system.nim (hunk 2 above)
Comment (Araq) [inline lib/system.nim]: Yeah, it's good enough the way it is now.

**Pull request closed without merging**